proxy/ptunnel/commands/tunnel.go

```diff
@@ -138,24 +182,69 @@ func (r *Tunnel) Run(cmd *cobra.Command, args []string) (err error) {
 	return err
 }
 
-func (r *Tunnel) RunListener(listener net.Listener, proxy *url.URL, host string, port uint16, cookie *http.Cookie, hostport string) error {
-	defer listener.Close()
-	for {
-		generic, err := listener.Accept()
-		if err != nil {
-			return err
+// listenerChan returns a channel that contains connections that the Listener is
+// accepting. If the listener encounters an error, it will fill in `e` before
+// closing the returned channel and terminating.
+func listenerChan(l net.Listener, e *error) <-chan net.Conn {
+	c := make(chan net.Conn)
+	go func() {
+		defer close(c)
+		for {
+			conn, err := l.Accept()
+			if err != nil {
+				*e = err
+				return
```

Comment:
errors from accept can be transient (example: not enough file descriptors), I don't think you should always exit the goroutine? Generally you want to keep the listener running in that case.

That said, the error is reported correctly, so I don't feel strongly about adding more complexity here.

Reply:
ack